internal/util/merge/merge3.go

```diff
@@ -163,3 +164,113 @@ func (p PruningLocalPackageReader) isExcluded(path string) bool {
 	}
 	return false
 }
+
+type ResourceMergeMatcher struct {
+	MergeOnPath bool
+}
+
+// IsSameResource determines if 2 resources are same to be merged by matching GKNN+filepath
+// Group, Kind are derived from resource metadata directly, Namespace and Name are derived
+// from merge comment which is of format "kpt-merge: namespace/name", if the merge comment
+// is not present, then it falls back to Namespace and Name on the resource meta
+func (rm *ResourceMergeMatcher) IsSameResource(node1, node2 *yaml.RNode) bool {
+	if node1 == nil || node2 == nil {
+		return false
+	}
+
+	meta1, err := node1.GetMeta()
+	if err != nil {
+		return false
+	}
+
+	meta2, err := node2.GetMeta()
+	if err != nil {
+		return false
+	}
+
+	if resolveGroup(meta1) != resolveGroup(meta2) {
+		return false
+	}
+
+	if meta1.Kind != meta2.Kind {
+		return false
+	}
+
+	if resolveName(meta1, metadataComment(node1)) != resolveName(meta2, metadataComment(node2)) {
+		return false
+	}
+
+	if resolveNamespace(meta1, metadataComment(node1)) != resolveNamespace(meta2, metadataComment(node2)) {
+		return false
+	}
+
+	if rm.MergeOnPath {
+		// directories may contain multiple copies of a resource with the same
+		// name, namespace, apiVersion and kind -- e.g. kustomize patches, or
+		// multiple environments
+		// mergeOnPath configures the merge logic to use the path as part of the
+		// resource key
+		if meta1.Annotations[kioutil.PathAnnotation] != meta2.Annotations[kioutil.PathAnnotation] {
+			return false
+		}
+	}
+	return true
+}
+
+// resolveGroup resolves the group of a resource from ResourceMeta
+func resolveGroup(meta yaml.ResourceMeta) string {
```

Comment:
I think there are libraries in kubernetes that does this. We should consider just using them.

Reply:
I din't find any. This is simple enough IMO.

Reply:
Kubernetes dependencies are painful. Used kyaml library instead. It has a util method too.